### PR TITLE
Fine tune _calculateTwap

### PR DESCRIPTION
**Pull request by twxia**
This PR is to fix an existing bug, https://github.com/perpetual-protocol/perp-oracle-contract/pull/62, that the TWAP is inaccurate in some conditions.

### CODE REVIEW

File: contracts/twap/CumulativeTwap.sol [hunk 1]
```diff
@@ -129,11 +147,15 @@ contract CumulativeTwap is BlockContext {
...
-            return (beforeOrAt, beforeOrAt);
+            // if the observation is the same as the targetTimestamp
+            // atOrAfter doesn't matter
+            // if the observation is less than the targetTimestamp
+            // atOrAfter repesents latest price and timestamp
+            return (beforeOrAt, atOrAfter);
         }
 
         // now, set before to the oldest observation
-        beforeOrAt = observations[(currentObservationIndex + 1) % MAX_OBSERVATION];
```
Comment (tienshaoku) [inline contracts/twap/CumulativeTwap.sol:136]: Need some help: why is this remainder removed?

Comment (tienshaoku) [reply contracts/twap/CumulativeTwap.sol:136]: Oh i just realised that there's a previous pr to this one: https://github.com/perpetual-protocol/perp-oracle-contract/pull/61
Will take a look at that and get back to this one

Comment (twxia) [reply contracts/twap/CumulativeTwap.sol]: you're right. I merge it with a wrong logic accidently.

File: contracts/twap/CumulativeTwap.sol [hunk 2]
```diff
@@ -91,33 +106,36 @@ contract CumulativeTwap is BlockContext {
...
         else {
-            uint256 observationTimeDelta = atOrAfter.timestamp - beforeOrAt.timestamp;
+            // it implies beforeOrAt = observations[currentObservationIndex]
+            // which means there's no atOrAfter
+            if (atOrAfter.timestamp == 0) {
+                atOrAfter.priceCumulative =
+                    beforeOrAt.priceCumulative +
+                    (beforeOrAt.price * (targetTimestamp - beforeOrAt.timestamp));
+                atOrAfter.timestamp = targetTimestamp;
+            }
```
Comment (paco0x) [inline contracts/twap/CumulativeTwap.sol:126]: we can directly return `targetCumulativePrice = atOrAfter.priceCumulative` here?

Comment (twxia) [reply contracts/twap/CumulativeTwap.sol:126]: ah right 🙏 thanks

Comment (twxia) [reply contracts/twap/CumulativeTwap.sol]: fixed in f64de2e

**Review by paco0x — APPROVED** [review]

**Comment by paco0x** [comment]
is it good to merge this PR? @twxia

**Pull request merged**